text/3453-f16-and-f128.md

```diff
@@ -0,0 +1,128 @@
+
+Crates may define their type on top of a C binding, but extended float type definition in C is complex and confusing. The meaning of C types may vary by target and/or compiler options. Implementing `f16` and `f128` in the Rust compiler helps to maintain a stable codegen interface and ensures that all users have one single canonical definition of 16-bit and 128-bit float types, making it easier to exchange data between crates and languages.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many other languages and compilers have support for these proposed float types. As mentioned above, C has `_Float16` and `_Float128`, and C++ has `std::float16_t` and `std::float128_t`. Glibc supports 128-bit floats in software on [many architectures](https://sourceware.org/git/?p=glibc.git;a=blob;f=NEWS;hb=81325b12b14c44887f1633a2c180a413afc2b504#l143).
+
+This RFC is designed as a subset of [RFC 3451](https://github.com/rust-lang/rfcs/pull/3451), which proposes adding a variety of float types, including ones not in this RFC designed for interoperability with other languages.
+
+Both this RFC and 3451 are built upon the discussion in [issue 2629](https://github.com/rust-lang/rfcs/issues/2629).
+
+The main consensus of the discussion thus far is that more float types would be useful, especially the IEEE 754 types proposed in this RFC as `f16` and `f128`. Other types can be discussed in a future RFC.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The main unresolved parts of this RFC are the implementation details in the context of the Rust compiler and standard library. The behavior of `f16` and `f128` is well-defined by the IEEE 754 standard, and is not up for debate. Whether these types should be included in the language is the main question of this RFC, which will be resolved when this RFC is accepted.
+
+Several future questions are intentionally left unresolved, and should be handled by another RFC. This RFC does not have the goal of covering the entire IEEE 754 standard, since it does not include `f256` and the decimal-float types. This RFC also does not have the goal of adding existing platform-specific float types such as x86's 80-bit double-extended-precision.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+See [RFC 3451](https://github.com/rust-lang/rfcs/pull/3451) for discussion about adding more float types. RFC 3451 is mostly a superset of this RFC.
+
+[^1]: Existing AI neural networks often use the [16-bit brain float format](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) instead of 16-bit half precision, which is a truncated version of 32-bit single precision. This is done to allow performing operations with 32-bit floats and quickly convert to 16-bit for storage. With modern architectures and languages supporting 16-bit half precision floats, newer AI neural networks may switch to half precision, which is more precise than brain float.
```

Comment:
Speculating that AI workloads will switch from bfloat16 to IEEE float16 is, well, speculation. 

The bfloat16 format was designed by people who already knew about the other format and didn't want to use it for their purposes (likely because the decrease in precision is offset by a huge increase in the maximum value going from **65,504** to **3 * 10<sup>38</sup>**)